### DISCUSSION TITLE
Questions about MapLayer with Google Map

### DISCUSSION THREAD

**Discussion by Myzik-Dev** [iTwin/itwinjs-core#2075] [Q&A] [answered]
In v2.16 Google Maps rendering worked well.
Google Maps rendering is not available in v2.17 and later.
It seems to be rendering with a Bingmap.

The TitleURL used for the source is as follows.
http://mt0.google.com/vt/lyrs=y&hl=en&x={column}&y={row}&z={level}&s=Ga

Is there a way to render Google Maps in v2.17 and later?

**Comment by mdastous-bentley** [comment, marked as answer]
Hi,

Assuming your iModel.js host application is https, you should also use a https base URL, so:
https://mt0.google.com/vt/lyrs=y&hl=en&x={column}&y={row}&z={level}&s=Ga

HTH,

Michel

**Reply by Myzik-Dev** [reply]
All services are using only http.
There are no source changes and Google Maps does not work only when using itwin.js v2.17 or later.

**Reply by mdastous-bentley** [reply]
I created this sandbox that uses Google Maps as background map:
[https://www.itwinjs.org/sandboxes/MichelDastous/google_maps_background](https://www.itwinjs.org/sandboxes/MichelDastous/google_maps_background)

Feel free to fork this sandbox to provide sample code that reproduce your issue.

HTH,

Michel

**Reply by Myzik-Dev** [reply]
Instead of the **'changeBackgroundMapProps**' I used,
It works well using the '**changeBaseMapProps**' you told me.

Thank you for your quick reply.

What is the difference between the two functions?

**Reply by mdastous-bentley** [reply]
Google Maps is not officially supported yet.  Please note changeBaseMapProps is marked as internal and might change in future versions.

**Reply by paprophet** [reply]
1.But when  i change to  tianditu map (https://www.tianditu.gov.cn ,   
url: "http://t0.tianditu.gov.cn/img_w/wmts?SERVICE=WMTS&REQUEST=GetTile&VERSION=1.0.0&LAYER=img&STYLE=default&TILEMATRIXSET=w&FORMAT=tiles&TILEMATRIX={level}&TILEROW={column}&TILECOL={row}&tk=19a0da10dec2f77233679367cc129658" )   ​
but  noting to   show and chrome not show error , i think there have coordinate transformation ？？？ 
2.
in "https://mt0.google.com/vt/lyrs=y&hl=en&x={column}&y={row}&z={level}&s=Ga" 
but how write the url like {longitude} or {latitude}?

**Reply by mdastous-bentley** [reply]
WMTS is not officially supported as a base map.  However, WMTS can be used to a map layer to you view.  We provide a widget for achieving this into the '@bentley/map-layers' package.

Google Maps is a tile format, and tile index(i.e. level, row, column) need to be provided in the URL in order to get a tile.  Are you trying to call an API that requires a specific position?

**Reply by Myzik-Dev** [reply]
> WMTS is not officially supported as a base map. However, WMTS can be used to a map layer to you view. We provide a widget for achieving this into the '@bentley/map-layers' package.
> 
> Google Maps is a tile format, and tile index(i.e. level, row, column) need to be provided in the URL in order to get a tile. Are you trying to call an API that requires a specific position?

As a result of the test, the resolution of the Korea area of ​​the map (due to the map-related laws) was the best, so I plan to use it.

**Reply by mdastous-bentley** [reply]
You can use the following line of code to use a WMTS base layer:
`viewState?.displayStyle.changeBaseMapProps({formatId:"WMTS", name: "WMTS_BaseMap", url: "<WMTS_endpoint>", subLayers:[{name: "<sublayer_name>", visible:true}]})`
**<WMTS_endpoint>**: Url to the WMTS endpoint, in your case should be ''http://t0.tianditu.gov.cn/img_w/wmts"
**<sublayer_name>**: Name of layer to publish.  The layer coordinate system must be EPSG:3857 and its tileset definition must be GoogleMaps compatible.

**Again, please note that changeBaseMapProps is still not published in API and is subject so change.**

**Reply by paprophet** [reply]
> You can use the following line of code to use a WMTS base layer:
> `viewState?.displayStyle.changeBaseMapProps({formatId:"WMTS", name: "WMTS_BaseMap", url: "<WMTS_endpoint>", subLayers:[{name: "<sublayer_name>", visible:true}]})`
> **<WMTS_endpoint>**: Url to the WMTS endpoint, in your case should be ''http://t0.tianditu.gov.cn/img_w/wmts"
> **<sublayer_name>**: Name of layer to publish. The layer coordinate system must be EPSG:3857 and its tileset definition must be GoogleMaps compatible.
> 
> **Again, please note that changeBaseMapProps is still not published in API and is subject so change.**

now the request url not contain the key for the map url 
![123](https://user-images.githubusercontent.com/18022720/131203801-2bcb23b4-9b15-4f39-948a-544a77737631.png)

**Reply by mdastous-bentley** [reply]
Hi,  can you please fill an issue for this, we will look into this. Thanks.

**Reply by gisjiashu** [reply]
Hi Myzik-Dev, 
For loading Tianditu map services, you can load it through tiles url services which in map-layers widget.
For example: http://t4.tianditu.com/DataServer?T=vec_w&x={column}&y={row}&l={level}&tk=xxxxxxx

**Reply by Myzik-Dev** [reply]
I don't know why you recommended me a Chinese map service (map.tianditu.gov.cn).
Korea and China are different countries.
As a result of the test, the resolution of the Korea area of ​​the map (due to the map-related laws) was the best.

My problem has been solved and well used with sandbox examples coded by 'mdastous-bentley'.

Thank you again.
